### DISCUSSION TITLE
Test implementation through the browser

### DISCUSSION THREAD

**Discussion by dbauszus-glx** [GEOLYTIX/xyz#1750] [Ideas]
I would like to discuss the implementation of browser tests.

I don't feel that having a test view with duplicate markup from the default view is the right approach.

I am playing with the idea of a testing plugin which can be built and provided in the public folder with the mapp and ui libraries.

The benefit of a testing plugin would be that the plugin can import the codi framework and other required esm libraries dynamically.

The plugin can be configured on the locale and receives the mapview and it's locale together with the plugin configuration itself.

This would allow to configure individual tests to be run. The plugin nature even allows for the testing configuration and execution to only be loaded and executed if requested with specific roles.

The testing plugin would also allow to provide an interface for running the tests.

This shouldn't be executed any time the view is loaded but it should be possible to trigger the execution through a test interface which may also provide a report where a development console may not be available.

**Comment by dbauszus-glx** [comment]
It is good to run a suite of many tests but it should also be possible to run an instance, hot rebuild, and run a specific test from the [plugin] test interface.

A plugin would also allow us to run specific tests on custom interfaces.

**Comment by RobAndrewHurst** [comment]
This is a good idea! Very keen to do away with the duplicate view as it doesn't provide much except maintenance.

**Comment by simon-leech** [comment]
Like the idea a lot, having an interface to only re-run a failing test rather than them all when making changes would be very useful. 

I also think if it's a plugin then it would be possible in the future to extend the tests to allow for custom methods passed from other resources such as bespoke plugins, custom views and so on too.
